sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -444,12 +444,22 @@ object ColumnPruning extends Rule[LogicalPlan] {
       f.copy(child = prunedChild(child, f.references))
     case e @ Expand(_, _, child) if (child.outputSet -- e.references).nonEmpty =>
       e.copy(child = prunedChild(child, e.references))
-    case g: Generate if !g.join && (g.child.outputSet -- g.references).nonEmpty =>
-      g.copy(child = prunedChild(g.child, g.references))
 
-    // Turn off `join` for Generate if no column from it's child is used
-    case p @ Project(_, g: Generate) if g.join && p.references.subsetOf(g.generatedSet) =>
-      p.copy(child = g.copy(join = false))
+    // prune unrequired references
+    case g : Generate
+      if (AttributeSet(g.unrequiredChildOutput) -- g.generator.references).nonEmpty =>
+        g.copy(child = prunedChild(g.child,
+                        g.child.outputSet -- g.unrequiredChildOutput ++ g.generator.references),
+               unrequiredChildOutput = g.generator.references.toSeq)
```

Comment:
This looks pretty complicated, how about
```
case g: Generate if (g.child.outputSet -- g.references -- g.requiredChildOutput).nonEmpty =>
  g.copy(child = prunedChild(g.child, g.references))
```
The logical is pretty simple, if child output has something that is not referred by generator, or by parent, prune the child.